py/desispec/scripts/qsoqn.py

```diff
@@ -277,17 +277,28 @@ def selection_targets_with_QN(redrock, fibermap, sel_to_QN, DESI_TARGET, spectra
                 "$DESI_ROOT is not set in the current environment. Please set it before running this code.")
             raise KeyError("QN_MODEL_FILE and DESI_ROOT are not set in the current environment.")
 
+    # Whether we should use linear weighting or not
+    # If a future QN weights file breaks the convention that a linear weights
+    # file should use "linear" in the name then this won't work...
+    if "linear" in model_QN_path:
+        linear_weights = True
+        wave_to_use = linear_wave
+    else: # logarithmic grid
+        linear_weights = False
+        wave_to_use = wave
+
     model_QN = load_model(model_QN_path)
```

Comment:
I'm a bit nervous about using the filename itself to make the linear vs. log determination.  Could that be derived from the contents of the file itself and an interface like:
```python
model_QN, wave_to_use = load_model(model_QN_path)
...
data, index_with_QN = load_desi_coadd(spectra_name, sel_to_QN, wave_to_use)
```

i.e. instead of passing around boolean flags and multiple wavelength grid options and deriving what to do from a filename, have the `load_model` tell you exactly which wavelength grid to use, and pass that into `load_desi_coadd` to enforce using that grid.  Would that work?

Reply:
I can make this happen but I should warn you that it would only be slightly less fragile while shifting the fragility into QuasarNP. Since the model output saved from tensorflow has no determination whether or not the input grid is logarithmic vs linear, the only way that QuasarNP would determine it within load_model would be to check whether or not the expected input dimensionality of the dataset is 443 (for logarithmic) or 458 (for linear). I'm mildly wary of this because it would be a significant api change that is backwards breaking (i.e. any old code that uses load_model will break with the additional return value) but if you think it's the best option I can implement it somewhat easily.

Reply:
Documenting a few items from a Slack conversation:
* let's post-process the new model file to include a new data element with the wavelengths used.  Code can then use that to know exactly what wavelengths to use without guessing.  If you don't want to touch the old model file, ok, the code could assume that if the "wavelengths" element is missing and the weights have size 443 then it is an old model file.
* desispec.zmtl is another place where QuasarNP is load and run, which needs to be updated
* Since Dylan + the DESI pipeline are the only known users of QuasarNP, I'm fine with breaking backwards compatibility of load_model to also return the wavelength array (since adapting to the change is also a simple end-user fix).  If we want to be really conservative, we could also add a new function `load_model_and_wavelengths` or a new option `load_model(filename, return_wavelengths=False)` that is backwards compatible but can opt-in to getting the wavelengths too.

Reply:
The updated QuasarNP is currently in a branch at https://github.com/desihub/QuasarNP/tree/auto_derive_linear, but can/will be pulled into the main branch and tagged. With this change:
- `load_model` now returns the wavelength grid used for the given weights file, auto assuming logarithmic grid unless specified within the weights file. qsoqn uses this grid for running QuasarNP. I've elected to just break backwards compat, we incremented the version number to 0.2.x anyway, if anyone needs backwards compatible with regards to return types and values API they can use 0.1.6. I'll make note of this in the next QuasarNP release notes. 
- There is a script in QuasarNP: [quasarnp/scripts/post_process_weights.py ](https://github.com/desihub/QuasarNP/blob/auto_derive_linear/scripts/post_process_weights.py) which can add wavelength information (linear or log) to weights files for loading with quasarnp

I've also made what I believe to be the requisite changes to zmtl.py, although I haven't had a chance yet to test and verify that it works as intended...